public/app/features/scenes/variables/VariableDependencyConfig.ts

```diff
@@ -0,0 +1,130 @@
+import { variableRegex } from 'app/features/variables/utils';
+
+import { SceneObject, SceneObjectState } from '../core/types';
+
+import { SceneVariable, SceneVariableDependencyConfigLike } from './types';
+
+interface VariableDependencyConfigOptions<TState extends SceneObjectState> {
+  /**
+   * State paths to scan / extract variable dependencies from. Leave empty to scan all paths.
+   */
+  statePaths?: Array<keyof TState>;
+  /**
+   * Optional way to customize how to handle when a dependent variable changes
+   * If not specified the default behavior is to trigger a re-render
+   */
+  onReferencedVariableValueChanged?: () => void;
+}
+
+export class VariableDependencyConfig<TState extends SceneObjectState> implements SceneVariableDependencyConfigLike {
+  private _state: TState | undefined;
+  private _dependencies = new Set<string>();
+  private _statePaths?: Array<keyof TState>;
+  private _onReferencedVariableValueChanged: () => void;
+
+  public scanCount = 0;
+
+  public constructor(private _sceneObject: SceneObject<TState>, options: VariableDependencyConfigOptions<TState>) {
+    this._statePaths = options.statePaths;
+    this._onReferencedVariableValueChanged =
+      options.onReferencedVariableValueChanged ?? this.defaultHandlerReferencedVariableValueChanged;
+  }
+
+  /**
+   * Used to check for dependency on a specific variable
+   */
+  public hasDependencyOn(name: string): boolean {
+    return this.getNames().has(name);
+  }
+
+  /**
+   * This is called whenever any set of variables have new values. It up to this implementation to check if it's relevant given the current dependencies.
+   */
+  public variableValuesChanged(variables: Set<SceneVariable>) {
+    const deps = this.getNames();
+
+    for (const variable of variables) {
+      if (deps.has(variable.state.name)) {
+        this._onReferencedVariableValueChanged();
+        return;
+      }
+    }
+  }
+
+  /**
+   * Only way to force a re-render is to update state right now
+   */
+  private defaultHandlerReferencedVariableValueChanged = () => {
+    this._sceneObject.setState({});
```

Comment:
Since we are using this both in this PR and in the Grid work - should we just extend the SceneObject interface with a `rerender` method to be more explicit?

Reply:
I think so, maybe call it forceRender?

Reply:
`forceRender` is fine IMO